extensions/chathistory.md

```diff
@@ -0,0 +1,115 @@
+
+If a nickname is given as the `target` then the server SHOULD include history sent between both the current user and the target nickname as to give a full conversation. The server SHOULD attempt to include history involving other nicknames if either the current user or the target nickname has changed during the requested timeframe.
+
+#### Subcommands
+
+The following subcommands are used to describe how the server should return messages relative to the `timestamp` or `msgid` given.
+
+#### `BEFORE`
+
+    CHATHISTORY BEFORE <target> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+    
+Request up to `limit` number of messages before and including the given `timestamp` or `msgid`. Only one timestamp or msgid MUST be given, not both.
+
+#### `AFTER`
+    CHATHISTORY AFTER <target> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+Request up to `limit` number of messages after and excluding the given `timestamp` or `msgid`. Only one timestamp or msgid MUST be given, not both.
+
+#### `LATEST`
+    CHATHISTORY LATEST <target> <* | timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+Request the most recent messages that have been sent after and excluding the given `timestamp` or `msgid`. If a `*` is given instead of a timestamp or msgid, the server MUST use the current time as a timestamp. The number of messages returned MUST be equal to or less than `limit`. If a `*` is not given, only one timestamp or msgid MUST be given, not both.
+
+This is useful for retrieving the latest conversation when first joining a channel or opening a query buffer.
+
+#### `AROUND`
+    CHATHISTORY AROUND <target> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+Request a number of messages before and after the `timestamp` or `msgid` with the total number of returned messages not exceeding `limit`. The implementation may decide how many messages to include before and after the selected message. Only one timestamp or msgid MUST be given, not both.
+
+This is useful for retrieving conversation context around a single message.
+
+#### `BETWEEN`
+    CHATHISTORY BETWEEN <target> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+Request up to `limit` number of messages between the given `timestamp` or `msgid` values. The returned messages MUST start from the inclusive first message selector, while excluding and finishing on the second - this may be forwards or backwards in time.
+
+#### Returned message notes
+The returned messages MUST be in ascending time order and the `server-time` tag SHOULD be the time at which the message was received by the IRC server. The `msgid` tag that identifies each individual message in a response MUST be the `msgid` tag as originally sent by the IRC server.
+
+#### Errors and Warnings
```

Comment:
Suggestions for more FAIL messages:

* `UNKNOWN_CRITERIA <criteria>` &mdash; for if someone specifies an invalid criteria like wibble=foo instead of e.g.  msgid=abc123. 

* `INVALID_CRITERIA <criteria>` &mdash; for if someone specifies a criteria thats not valid for a command e.g. the wildcard (`*`) criteria with a non-LATEST command.

* `INVALID_LIMIT <limit>` &mdash; for if someone specifies a limit that is out of the allowed range.